src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/TagHelperSemanticRangeVisitor.cs

```diff
@@ -178,39 +290,93 @@ private bool ClassifyTagName(MarkupTagHelperElementSyntax node)
             return false;
         }
 
-        private SemanticRange CreateSemanticRange(SyntaxNode node, SyntaxKind kind)
+        private void AddSemanticRange(SyntaxNode node, int semanticKind)
         {
+            var source = _razorCodeDocument.Source;
+            var range = node.GetRange(source);
+
+            var semanticRange = new SemanticRange(semanticKind, range);
+
+            if (_range is null || semanticRange.Range.OverlapsWith(_range))
+            {
+                _semanticRanges.Add(semanticRange);
+            }
+        }
+
+        private void AddSemanticRange(SyntaxNode node, SyntaxKind? kind = null)
+        {
+            if(node is null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            if (kind is null)
+            {
+                kind = node.Kind;
+            }
+
             int semanticKind;
             switch (kind)
             {
                 case SyntaxKind.MarkupTagHelperDirectiveAttribute:
                 case SyntaxKind.MarkupMinimizedTagHelperDirectiveAttribute:
-                    semanticKind = RazorSemanticTokensLegend.TokenTypesLegend[RazorSemanticTokensLegend.RazorDirectiveAttribute];
+                    semanticKind = RazorSemanticTokensLegend.RazorDirectiveAttribute;
                     break;
                 case SyntaxKind.MarkupTagHelperStartTag:
                 case SyntaxKind.MarkupTagHelperEndTag:
-                    semanticKind = RazorSemanticTokensLegend.TokenTypesLegend[RazorSemanticTokensLegend.RazorTagHelperElement];
+                    semanticKind = RazorSemanticTokensLegend.RazorTagHelperElement;
                     break;
                 case SyntaxKind.MarkupTagHelperAttribute:
                 case SyntaxKind.MarkupMinimizedTagHelperAttribute:
-                    semanticKind = RazorSemanticTokensLegend.TokenTypesLegend[RazorSemanticTokensLegend.RazorTagHelperAttribute];
+                    semanticKind = RazorSemanticTokensLegend.RazorTagHelperAttribute;
                     break;
                 case SyntaxKind.Transition:
-                    semanticKind = RazorSemanticTokensLegend.TokenTypesLegend[RazorSemanticTokensLegend.RazorTransition];
+                    semanticKind = RazorSemanticTokensLegend.RazorTransition;
                     break;
                 case SyntaxKind.Colon:
-                    semanticKind = RazorSemanticTokensLegend.TokenTypesLegend[RazorSemanticTokensLegend.RazorDirectiveColon];
+                    semanticKind = RazorSemanticTokensLegend.RazorDirectiveColon;
+                    break;
+                case SyntaxKind.RazorDirective:
+                    semanticKind = RazorSemanticTokensLegend.RazorDirective;
+                    break;
+                case SyntaxKind.RazorCommentTransition:
+                    semanticKind = RazorSemanticTokensLegend.RazorCommentTransition;
```

Comment:
I have some doubts about having such verbosity in the various token kinds vs. just having a "metacode" semantic kind instead all the various comment transition pieces. Ditto with some of the HTML concepts; that being said though, this is your baby so we can see how it plays out 😄

Reply:
I split the `@*`'s into @ and * because 1. That's how our current VSCode TextMate grammar does it (hard to tell what exactly VS does for those tokens, they're both listed as "HTML Server-Side Script", but I don't know if they returned one token or two) 2. It would have required some changes to AddSemanticRange to combine multiple nodes into one Semantic Tokens range. #2 isn't that big of a deal, so if we don't like it I can make the change, I just didn't think it would matter in practice.

Reply:
> 2 isn't that big of a deal, so if we don't like it I can make the change, I just didn't think it would matter in practice.

Up to you! Oh and I wasn't referring to merging tokens together. I was just talking about having less semantic kinds. That being said, I don't feel strongly.

Reply:
Oh chill, I did end up consolidating semantic kinds a bit so hopefully it's more pleasing now.